Test1/ConcurrentPriorityQueue.cs

```diff
@@ -0,0 +1,51 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Linq;
+using System.Threading;
+
+namespace Test1;
+
+public class ConcurrentPriorityQueue<TP, TV> : IConcurrentPriorityQueue<TP, TV> where TP : IComparable<TP>
+{
+    private readonly SortedDictionary<TP, List<TV>> _elementsByPriority = new();
+    private uint _elementsCount = 0;
+    private readonly object _elementsLock = new();
+
+    public void Enqueue(TP priority, TV value)
+    {
+        lock (_elementsLock)
+        {
+            if (!_elementsByPriority.ContainsKey(priority))
+            {
+                _elementsByPriority[priority] = new List<TV>();
+            }
+
+            _elementsByPriority[priority].Add(value);
+            _elementsCount++;
+
+            Monitor.PulseAll(_elementsLock);
+        }
+    }
+
+    public TV Dequeue()
+    {
+        lock (_elementsLock)
+        {
+            if (_elementsCount == 0) Monitor.Wait(_elementsLock);
```

Comment:
```suggestion
            while (_elementsCount == 0) Monitor.Wait(_elementsLock);
```
А если несколько потоков заблокируются на `Wait`? Затем в очередь добавят элемент, все эти потоки проснутся. Один из них возьмет элемент из очереди, а остальные упадут на `queue.Keys.Last()`. 
Я ведь говорил на паре, что не `if`, а `while` тут нужно использовать.

Reply:
Справедливо

Reply:
Тогда разумней использовать Pulse() вместо PulseAll()?

Reply:
Но у меня используется Pulse() вместо PulseAll(), поэтому такой ситуации не возникнет

Reply:
Даже с `Pulse()` нужен `while`. На паре обсудили.
-1